spec/new_integration/rack_app/rack_app_spec.rb

```diff
@@ -100,10 +122,256 @@ def handle(*, res)
       expect(last_response.status).to eq 200
       expect(last_response.body).to eq "Hello world"
 
+      get "/inline"
+
+      expect(last_response.status).to eq 200
+      expect(last_response.body).to eq "Inline"
+
+      get "/health"
+
+      expect(last_response.status).to eq 200
+      expect(last_response.body).to eq "Health, OK"
+
       get "/admin/dashboard"
 
       expect(last_response.status).to eq 200
       expect(last_response.body).to eq "Admin dashboard here"
     end
   end
+
+  specify "It gives priority to app actions" do
+    with_tmp_directory(Dir.mktmpdir) do
+      write "config/application.rb", <<~RUBY
+        require "hanami"
+
+        module TestApp
+          class Application < Hanami::Application
+            config.logger.stream = File.new("/dev/null", "w")
+          end
+        end
+      RUBY
+
+      write "config/routes.rb", <<~RUBY
+        module TestApp
+          class Routes < Hanami::Routes
+            define do
+              root to: "home.index"
+
+              slice :main, at: "/" do
+                root to: "home.index"
+              end
+            end
+          end
+        end
+      RUBY
+
+      write "app/actions/home/index.rb", <<~RUBY
+        require "hanami/action"
+
+        module TestApp
+          module Actions
+            module Home
+              class Index < Hanami::Action
+                def handle(*, res)
+                  res.body = "Hello from App"
+                end
+              end
+            end
+          end
+        end
+      RUBY
+
+      write "slices/main/actions/home/index.rb", <<~RUBY
+        require "hanami/action"
+
+        module Main
+          module Actions
+            module Home
+              class Index < Hanami::Action
+                def handle(*, res)
+                  res.body = "Hello from Slice"
+                end
+              end
+            end
+          end
+        end
+      RUBY
+
+      require "hanami/boot"
+
+      get "/"
+
+      expect(last_response.status).to eq 200
+      expect(last_response.body).to eq "Hello from App"
+    end
+  end
+
+  xspecify "It does not choose actions from slice to route in app" do
```

Comment:
@jodosha I've fixed this one: it was an unrelated failure due to the API module being named `Api`, instead of our expected `API` (dry-inflector has `API` in its list of default abbreviations).